File: src/main/java/puzzles/year24/day1/Part2.java
package puzzles.year24.day1;

import util.Puzzle;

import java.util.ArrayList;
import java.util.List;

public class Part2 implements Puzzle<Integer>  {
    @Override
    public Integer getAnswer(String inputString) {
        String[] lines = inputString.split("\r\n");
        List<Integer> first = new ArrayList<>();
        List<Integer> second = new ArrayList<>();

        for (String line : lines) {
            String[] pair = line.split(" {3}");
            first.add(Integer.parseInt(pair[0]));
            second.add(Integer.parseInt(pair[1]));
        }

        first.sort(Integer::compareTo);
        second.sort(Integer::compareTo);

        int result = 0;
        for (Integer i : first) {
            int count = 0;
            boolean found = false;
            for (Integer j : second) {
                if (i.equals(j)) {
                    count++;
                    found = true;
                } else if (found) {
                    break;
                }
            }
            result += count * i;
        }

        return result;
    }
}

File: src/main/java/puzzles/year25/day10/Part2.kt
package puzzles.year25.day10

import org.chocosolver.solver.Model
import org.chocosolver.solver.Solver
import util.Puzzle

class Part2 : Puzzle<Int?> {
    override fun getAnswer(inputString: String): Int {
        val input = inputString.trim().lines().map { it.split(" ") }.map { parse(it) }
        var result = 0
        for (i in input) {
            val matrix = calculateMatrix(i)
            result += linearSolver(matrix, i.second.size, i.first.max(), i.first.min(), (i.first.sum() / 1.5).toInt())
        }

        return result
    }

    fun linearSolver(input: Pair<Array<Array<Int>>, Array<Int>>, buttons: Int, upperBound: Int, solutionLowerBound: Int, solutionUpperBound: Int): Int {
        val model = Model("Model")

        val vars = model.intVarArray("x", buttons, 0, upperBound)

        for (i in input.first.indices) {
            val coefficients = input.first[i].toIntArray()
            val rhs = input.second[i]

            model.scalar(vars, coefficients, "=", rhs).post()
        }

        val objectiveSum = model.intVar("objectiveSum", solutionLowerBound, solutionUpperBound)
        model.sum(vars, "=", objectiveSum).post()

        val solver: Solver = model.solver

        val solution = solver.findOptimalSolution(objectiveSum, false)

        if (solution != null) {
            println(vars.contentToString())

            return solution.getIntVal(objectiveSum)
        } else {
            println("No integer solution found.")
            return 0
        }
    }

    fun parse(input: List<String>): Pair<List<Int>, List<List<Int>>> {
        var jolts: List<Int> = ArrayList()
        val buttons: MutableList<List<Int>> = ArrayList()

        for (s in input) {
            if (s.startsWith("{")) {
                jolts = s.substring(1, s.length - 1).split(",").map { it.toInt() }
            } else if (s.startsWith("(")) {
                buttons.add(s.substring(1, s.length - 1).split(",").map { it.toInt() })
            }
        }

        return Pair(jolts, buttons)
    }

    fun calculateMatrix(input: Pair<List<Int>, List<List<Int>>>): Pair<Array<Array<Int>>, Array<Int>> {
        val buttons = input.second
        val jolts = input.first
        val result = Array(jolts.size) { Array(buttons.size) { 0 } }
        val constants = jolts.toTypedArray()

        // Simple jolt equations
        for (j in jolts.indices) {
            for (b in buttons.indices) {
                if (buttons[b].contains(j)) {
                    result[j][b] = 1
                }
            }
        }

        return Pair(result, constants)
    }
}

// 18231 -- too low